Moved expires to resource metadata for services.Users.

russjones opened the pull request:
**Description**

Moved expiry field from spec to metadata for services.Users and updated
expiry check to prefer metadata and fallback to spec if not found. Added
test coverage.

russjones committed:
Moved expires to resource metadata for services.Users.

Moved expiry field from spec to metadata for services.Users and updated
expiry check to prefer metadata and fallback to spec if not found. Added
test coverage.

r0mant commented on lib/services/suite/suite.go:
```diff
@@ -41,6 +42,8 @@ import (
 	"gopkg.in/check.v1"
 )
 
+var _ = fmt.Printf
```
Need this?

r0mant commented on lib/auth/saml_test.go:
```diff
@@ -0,0 +1,112 @@
+	c clockwork.FakeClock
+}
+
+var _ = fmt.Printf
```
Need this?

klizhentas commented:
I think this problem applies to all users, not just github ones

klizhentas reviewed (changes requested):
Can you please check that if this problem applies to all external types

russjones commented:
@klizhentas It does and I fixed it for all external identity providers. I've fixed it for all and covered all with tests. That's what the bulk of the PR was.

russjones commented:
@gravitational-jenkins retest this please

klizhentas commented:
k, sorry the PR only showed me two github files, let me re-review

klizhentas commented:
retest this please

russjones committed:
Fixed TestStreamEvents.

russjones committed:
Consolidate external identity user creation code.

klizhentas commented on lib/auth/github.go:
```diff
@@ -210,61 +203,157 @@ func (s *AuthServer) validateGithubAuthCallback(q url.Values) (*GithubAuthRespon
+	return certs.ssh, certs.tls, nil
+}
+
+// createUserParams is a set of parameters used to create a user for an\
```
extra `\`?

klizhentas commented on lib/auth/github.go:
```diff
@@ -210,61 +203,157 @@ func (s *AuthServer) validateGithubAuthCallback(q url.Values) (*GithubAuthRespon
+}
+
+func (s *AuthServer) calculateGithubUser(connector services.GithubConnector, claims *services.GithubClaims, request *services.GithubAuthRequest) (*createUserParams, error) {
+	var p createUserParams
```
why not:

```go
p := createUserParams{
     connectorName: connector.GetName(),
     username: claims.Username,
}
```

klizhentas commented on lib/auth/github.go:
```diff
@@ -210,61 +203,157 @@ func (s *AuthServer) validateGithubAuthCallback(q url.Values) (*GithubAuthRespon
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	roleTTL := roles.AdjustSessionTTL(defaults.CertDuration)
```
should be Max duration, not default cert duration, so we can process larger than default values

klizhentas commented on lib/auth/oidc.go:
```diff
@@ -325,52 +314,71 @@ func claimsToTraitMap(claims jose.Claims) map[string][]string {
-	roles, err := a.buildOIDCRoles(connector, claims)
+func (a *AuthServer) calculateOIDCUser(connector services.OIDCConnector, claims jose.Claims, ident *oidc.Identity, request *services.OIDCAuthRequest) (*createUserParams, error) {
+	var err error
+	var p createUserParams
```
ditto as other user

klizhentas commented on lib/auth/saml.go:
```diff
@@ -105,47 +120,74 @@ func assertionsToTraitMap(assertionInfo saml2.AssertionInfo) map[string][]string
-	roles, err := a.buildSAMLRoles(connector, assertionInfo)
+func (a *AuthServer) calculateSAMLUser(connector services.SAMLConnector, assertionInfo saml2.AssertionInfo, request *services.SAMLAuthRequest) (*createUserParams, error) {
+	var err error
+	var p createUserParams
```
ditto

klizhentas reviewed (changes requested):
still have a couple of comments and one actual change
